### PR TITLE
Kubeadm patches: use config for newer versions; --experimental-patches for older

### PR DESCRIPTION
**Pull request by pacoxu**
https://github.com/kubernetes/kubeadm/tree/master/kinder/pkg/cluster/manager/actions

use --config for "init|join" and --patches for "upgrade"
for https://github.com/kubernetes/kubeadm/issues/2046 and https://github.com/kubernetes/kubernetes/pull/104065

### CODE REVIEW

File: kinder/pkg/kubeadm/discovery.go [hunk 1]
```diff
@@ -93,6 +93,39 @@ discovery:
   file:
     kubeConfigPath: %s`
 
+// GetPatchesDirectory returns the kubeadm config patch that will instruct kubeadm
+// to use patches directory.
+func GetPatchesDirectory(kubeadmConfigVersion string) (string, error) {
+	// select the patches for the kubeadm config version
+	log.Debugf("Preparing patchesDirectoryPatch for kubeadm config %s", kubeadmConfigVersion)
```
Comment (neolit123) [inline kinder/pkg/kubeadm/discovery.go:100]:
```suggestion
	log.Debugf("Preparing patches directory for kubeadm config %s", kubeadmConfigVersion)
```

File: kinder/pkg/kubeadm/discovery.go [hunk 2]
```diff
@@ -93,6 +93,39 @@ discovery:
...
+
+	var patch string
+	switch kubeadmConfigVersion {
+	case "v1beta2":
+		patch = patchesDirectoryPatchv1beta2
```
Comment (neolit123) [inline kinder/pkg/kubeadm/discovery.go:105]: [1] should return error as patches are not supported in v1beta2.

suggesting:
```
	switch {
	case "v1beta3":
		patch = patchesDirectoryPatchv1beta3
	default:
		return "", errors.Errorf("unsupported kubeadm config version for patches: %s", kubeadmConfigVersion)
	}
```

File: kinder/pkg/kubeadm/discovery.go [hunk 3]
```diff
@@ -93,6 +93,39 @@ discovery:
...
+
+	return fmt.Sprintf(patch, constants.PatchesDir), nil
+}
+
+const patchesDirectoryPatchv1beta2 = `apiVersion: kubeadm.k8s.io/v1beta2
+kind: JoinConfiguration
+metadata:
+  name: config
+patches:
+  directory: %s`
```
Comment (neolit123) [inline kinder/pkg/kubeadm/discovery.go:120]: we can remove this one.

File: kinder/pkg/cluster/manager/actions/kubeadm-config.go [hunk 4]
```diff
@@ -276,6 +276,13 @@ func getKubeadmConfig(c *status.Cluster, n *status.Node, data kubeadm.ConfigData
 		}
 		patches = append(patches, fileDiscoveryPatch)
 
+		// add patches directory to the config
+		patchesDirectoryPatch, err := kubeadm.GetPatchesDirectory(kubeadmConfigVersion)
+		if err != nil {
+			return "", err
+		}
+		patches = append(patches, patchesDirectoryPatch)
```
Comment (neolit123) [inline kinder/pkg/cluster/manager/actions/kubeadm-config.go:284]:
```suggestion
		patchesDirectory, err := kubeadm.GetPatchesDirectory(kubeadmConfigVersion)
		if err != nil {
			return "", err
		}
		patches = append(patches, patchesDirectory)
```

File: kinder/pkg/cluster/manager/actions/kubeadm-init.go [hunk 5]
```diff
@@ -101,9 +101,6 @@ func kubeadmInit(cp1 *status.Node, copyCertsMode CopyCertsMode, patchesDir, igno
 			// NB. certificate key is passed via the config file)
 		)
 	}
-	if patchesDir != "" {
-		initArgs = append(initArgs, "--experimental-patches", constants.PatchesDir)
-	}
```
Comment (neolit123) [inline kinder/pkg/cluster/manager/actions/kubeadm-init.go:106]: for a period of time we will have tests for 1.20, 1.21, 1.22, 1.23

this means:
- 1.20, 1.21, must use --experimental-patches
- "init"/"join" for 1.22, 1.23 can use v1beta3 / patches from config only.
- "upgrade" for 1.22, 1.23 must use --patches

so it's a bit trickier.

see for example on how to branch on versions:
https://github.com/kubernetes/kubeadm/blob/a29017a17040d7dd2f10eac92647e9aaed26c300/kinder/pkg/cluster/manager/actions/kubeadm-init.go#L44-L51

File: kinder/pkg/kubeadm/discovery.go [hunk 6]
```diff
@@ -93,6 +93,39 @@ discovery:
   file:
     kubeConfigPath: %s`
 
+// GetPatchesDirectory returns the kubeadm config patch that will instruct kubeadm
+// to use patches directory.
+func GetPatchesDirectory(kubeadmConfigVersion string) (string, error) {
```
Comment (neolit123) [inline kinder/pkg/kubeadm/discovery.go:98]:
```suggestion
func GetPatchesDirectoryPatch(kubeadmConfigVersion string) (string, error) {
```

File: kinder/pkg/cluster/manager/actions/kubeadm-init.go (hunk 5 above)
Comment (pacoxu) [reply kinder/pkg/cluster/manager/actions/kubeadm-init.go]: I updated the PR to check the version before applying --patch.

File: kinder/pkg/cluster/manager/actions/kubeadm-init.go [hunk 7]
```diff
@@ -101,8 +101,14 @@ func kubeadmInit(cp1 *status.Node, copyCertsMode CopyCertsMode, patchesDir, igno
 			// NB. certificate key is passed via the config file)
 		)
 	}
+
 	if patchesDir != "" {
-		initArgs = append(initArgs, "--experimental-patches", constants.PatchesDir)
+		if cp1.MustKubeadmVersion().LessThan(constants.V1_21) {
```
Comment (neolit123) [inline kinder/pkg/cluster/manager/actions/kubeadm-init.go:106]: should be <1.22 (here and in other places)
we are adding the option from config in 1.22.

File: kinder/pkg/cluster/manager/actions/kubeadm-upgrade.go [hunk 8]
```diff
@@ -131,7 +131,7 @@ func kubeadmUpgradeApply(c *status.Cluster, cp1 *status.Node, upgradeVersion *K8
 		"upgrade", "apply", "-f", fmt.Sprintf("v%s", upgradeVersion), fmt.Sprintf("--v=%d", vLevel),
 	}
 	if patchesDir != "" {
-		applyArgs = append(applyArgs, "--experimental-patches", constants.PatchesDir)
+		applyArgs = append(applyArgs, "--patches", constants.PatchesDir)
```
Comment (neolit123) [inline kinder/pkg/cluster/manager/actions/kubeadm-upgrade.go:134]: we also need the <1.22 version check in this function and kubeadmUpgradeNode

File: kinder/pkg/cluster/manager/actions/kubeadm-init.go [hunk 9]
```diff
@@ -101,8 +101,14 @@ func kubeadmInit(cp1 *status.Node, copyCertsMode CopyCertsMode, patchesDir, igno
...
+			initArgs = append(initArgs, "--experimental-patches", constants.PatchesDir)
+		}
+		if err := copyPatchesToNode(cp1, patchesDir); err != nil {
```
Comment (neolit123) [inline kinder/pkg/cluster/manager/actions/kubeadm-init.go:109]: this is already being called here:
https://github.com/kubernetes/kubeadm/blob/ebd1ae0b04a22e94bc73d07f10a5c2416ac7afe0/kinder/pkg/cluster/manager/actions/kubeadm-init.go#L43-L51

i think we can remove the V1_19 constant and the code above.

File: kinder/pkg/cluster/manager/actions/kubeadm-join.go [hunk 10]
```diff
@@ -139,9 +140,10 @@ func kubeadmJoinControlPlaneWithPhases(cp *status.Node, patchesDir, ignorePrefli
 		fmt.Sprintf("--config=%s", constants.KubeadmConfigPath),
 		fmt.Sprintf("--v=%d", vLevel),
 	}
-
```
Comment (neolit123) [inline kinder/pkg/cluster/manager/actions/kubeadm-join.go:142]: in general, we should avoid whitespace changes like adding lines and removing lines as part of other code changes.

File: kinder/pkg/cluster/manager/actions/kubeadm-join.go [hunk 11]
```diff
@@ -106,9 +106,10 @@ func kubeadmJoinControlPlane(cp *status.Node, patchesDir, ignorePreflightErrors
 		fmt.Sprintf("--v=%d", vLevel),
 	}
 	if patchesDir != "" {
-		joinArgs = append(joinArgs, "--experimental-patches", constants.PatchesDir)
+		if cp.MustKubeadmVersion().LessThan(constants.V1_21) {
+			joinArgs = append(joinArgs, "--experimental-patches", constants.PatchesDir)
+		}
```
Comment (neolit123) [inline kinder/pkg/cluster/manager/actions/kubeadm-join.go:111]: i think we can remove this check above:
https://github.com/kubernetes/kubeadm/blob/47c0def8aff991e23c599e207b66edb91632c4fa/kinder/pkg/cluster/manager/actions/kubeadm-join.go#L46-L54

and add copyPatchesToNode here.

File: kinder/pkg/constants/constants.go [hunk 12]
```diff
@@ -120,6 +120,8 @@ const (
 var (
 	// V1.19 minor version
 	V1_19 = K8sVersion.MustParseSemantic("v1.19.0-0")
+	// V1.21 minor version
+	V1_21 = K8sVersion.MustParseSemantic("v1.21.0-0")
```
Comment (neolit123) [inline kinder/pkg/constants/constants.go:124]:
```suggestion
	V1_21 = K8sVersion.MustParseSemantic("v1.22.0-0")
```

File: kinder/pkg/cluster/manager/actions/kubeadm-upgrade.go [hunk 13]
```diff
@@ -131,7 +128,9 @@ func kubeadmUpgradeApply(c *status.Cluster, cp1 *status.Node, upgradeVersion *K8
 		"upgrade", "apply", "-f", fmt.Sprintf("v%s", upgradeVersion), fmt.Sprintf("--v=%d", vLevel),
 	}
 	if patchesDir != "" {
-		applyArgs = append(applyArgs, "--experimental-patches", constants.PatchesDir)
+		if cp1.MustKubeadmVersion().LessThan(constants.V1_22) {
+			applyArgs = append(applyArgs, "--patches", constants.PatchesDir)
```
Comment (neolit123) [inline kinder/pkg/cluster/manager/actions/kubeadm-upgrade.go:132]:
```suggestion
			applyArgs = append(applyArgs, "--experimental-patches", constants.PatchesDir)
```

should be the old flag for < 1.22

File: kinder/pkg/cluster/manager/actions/kubeadm-upgrade.go [hunk 14]
```diff
@@ -159,7 +158,7 @@ func kubeadmUpgradeNode(c *status.Cluster, n *status.Node, upgradeVersion *K8sVe
 		"upgrade", "node", fmt.Sprintf("--v=%d", vLevel),
 	}
 	if patchesDir != "" {
-		nodeArgs = append(nodeArgs, fmt.Sprintf("--experimental-patches=%s", constants.PatchesDir))
+		nodeArgs = append(nodeArgs, fmt.Sprintf("--patches=%s", constants.PatchesDir))
```
Comment (neolit123) [inline kinder/pkg/cluster/manager/actions/kubeadm-upgrade.go:161]: same here. should be:

```
if v < 1.22 {
// use --experimental-patches
} else {
// use --patches
} 
```

**Comment by pacoxu** [comment]
Just got some time to update.

**Review by neolit123 — COMMENTED** [review]
/lgtm
/approve

thanks @pacoxu , this seems fine.

**Comment by neolit123** [comment]
/hold
please squash the commits, though.

**Comment by pacoxu** [comment]
Squashed.

**Pull request merged**